### PR TITLE
feat(select): add infinite scroll callback

### PR DESCRIPTION
**Pull request by grabkowski**
### Proposed behaviour

Screenshot 2020-11-26 at 18 13 15

Add a callback that is triggered when a user scrolls to the bottom of the list, to allow infinite scroll implementation.
Add an animation when the list is being resized.

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions
1. npm start
2. http://localhost:9001/?path=/docs/design-system-select--with-infinite-scroll
3. http://localhost:9001/?path=/docs/design-system-select-filterable--with-infinite-scroll

### CODE REVIEW

**Review by jamime — COMMENTED** [review]
When using the keyboard to navigate, when you get to the bottom of the list the selection moves to the first option so the user won't be able to see that more items are loading.

I think we should disable the behaviour when users have `onListScrollBottom` set, but add a new prop so they can enable the looping once all of the results have been loaded.

**Review by edleeks87 — COMMENTED** [review]
The function names are quite intuitive but I think it will be helpful to have some additional comments for the more complex parts as well; it will help others maintain the code and hopefully prevent the introduction of bugs and regressions.

It would also be worth adding an example of how consumers might implement the failed loading scenario IMO.

File: src/components/select/select-list/select-list.component.js [hunk 1]
```diff
@@ -40,65 +37,86 @@ const SelectList = React.forwardRef(
...
 
-    const childOptions = useMemo(() => React.Children.toArray(children), [
+    const childrenList = useMemo(() => React.Children.toArray(children), [
       children,
     ]);
 
+    const lastOptionIndex = useMemo(() => {
+      let lastIndex = 0;
+
+      childrenList.forEach((element, index) => {
```
[the remark this answers is not in the document]
Comment (grabkowski) [reply src/components/select/select-list/select-list.component.js:58]: That will just reverse the original array, I could clone it and then reverse it, but I will still need the correct index of the last element so then I'll have to subtract that index from the array length.
I think it's better to just loop through that array as it's not doing anything resource consuming on each loop and we have a bit cleaner code.

**Review by jamime — DISMISSED** [review]

File: src/components/select/select-list/select-list.component.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/select/select-list/select-list.component.js:58]: Ah yeah that's a good point, I agree it's best to leave it as is then

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — DISMISSED** [review]

File: cypress/features/regression/designSystem/filterableSelect.feature [hunk 2]
```diff
@@ -24,42 +23,54 @@ Feature: Design System Filterable Select component
...
+      And I type "Amber" into input
     When I click on "first" option on Select list
     Then Design system Select input has "Amber" value
       And "filterable" Select list is closed
 
   @positive
   Scenario: Selecting the action button opens up a dialog box
-    Given I open "Design System Select filterable" component page "with_action_button" in no iframe
+    Given I open "Design System Select filterable" component page "controlled" in no iframe
+      And I open "Design System Select filterable" component page "with_action_button" in no iframe
```
[the remark this answers is not in the document]
Comment (KatarzynaQA) [reply cypress/features/regression/designSystem/filterableSelect.feature]: fixed

File: cypress/features/regression/designSystem/simpleSelect.feature [hunk 3]
```diff
@@ -50,12 +53,27 @@ Feature: Design System Select component
...
+
+  @positive
+  Scenario: Lazy loading is visible after open the Simple Select
+    Given I open "Design System Select" component page "with_is_loading_prop" in no iframe
+    When I click on Select input with lazy loading in no iframe
+    Then Lazy loading is visible
+
+  @positive
+  Scenario: Lazy loading is visible after open and scroll to the botton of the Filterable Select
+    Given I open "Design System Select filterable" component page "with infinite scroll" in no iframe
```
Comment (DlgSHi) [inline cypress/features/regression/designSystem/simpleSelect.feature:73]: `filterable` or `simple` select should be?

**Review by DlgSHi — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — DISMISSED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by jamime — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Pull request merged**